server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -63,12 +64,28 @@ public class DesiredBalanceComputer {
         Setting.Property.NodeScope
     );
 
+    public static final Setting<TimeValue> MAX_BALANCE_COMPUTATION_TIME_DURING_INDEX_CREATION_SETTING = Setting.timeSetting(
+        "cluster.routing.allocation.desired_balance.max_balance_computation_time_during_index_creation",
+        TimeValue.timeValueSeconds(10),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
     private TimeValue progressLogInterval;
+    private long maxBalanceComputationTimeDuringIndexCreationMillis;
 
     public DesiredBalanceComputer(ClusterSettings clusterSettings, ThreadPool threadPool, ShardsAllocator delegateAllocator) {
```

Comment:
In a follow up, I'll just remove this constructor and replace its usages. Here it would add too much noise.

Reply:
:+1: I wish we have a dedicated interface for supplying time.
Often it is not clear if we actually submit tasks or just need a time.

Reply:
Agreed. If we're going to use a `LongSupplier` however, I think it should specify somewhere in the name what unit it's in (e.g. `timeSupplierInMilliseconds`), but a dedicated interface would be much better.

Reply:
Opened https://github.com/elastic/elasticsearch/pull/116333.